File: BEmu/ReferenceDataRequest/EventReference.cs
//------------------------------------------------------------------------------
// <copyright project="BEmu" file="ReferenceDataRequest/EventReference.cs" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

namespace BEmu.ReferenceDataRequest
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;
    using System.Text.RegularExpressions;

    internal class EventReference : Event
    {
        private readonly List<Message> _messages;

        internal EventReference(RequestReference request)
        {
            base._request = request;
            this._messages = this.GenerateMessages();
        }

        public override IEnumerable<Message> GetMessages()
        {
            return this._messages;
        }

        private List<Message> GenerateMessages()
        {
            List<Message> result = new List<Message>();
            var rreq = (RequestReference)base._request;

            var securities = new Dictionary<string, Dictionary<string, object>>();
            foreach (var security in rreq.Securities)
            {
                if (!securities.ContainsKey(security))
                {
                    bool isOption = Regex.IsMatch(security.ToUpper(), @"[A-Z]{1,4}\s+\d{6}[CP]\d{8} EQUITY"); //options look like: AAPL 150117C00600000 EQUITY

                    var fieldData = new Dictionary<string, object>();
                    securities.Add(security, fieldData);

                    List<string> badFields = new List<string>();
                    foreach (var field in rreq.Fields)
                    {
                        object value = Types.RandomDataGenerator.ReferenceDataFromFieldName(field, security, isOption, rreq);
                        if (value != null)
                        {
                            fieldData.Add(field.ToUpper(), value);
                        }
                    }
                }
            }

            MessageReference msg = new MessageReference(base._request.correlationId, securities);
            result.Add(msg);

            return result;
        }
    }
}

File: BEmuJava/bemu/src/main/java/com/bemu/BEmu/ReferenceDataRequest/EventReference.java
package com.bemu.BEmu.ReferenceDataRequest;

import com.bemu.BEmu.Event;
import com.bemu.BEmu.Message;
import com.bemu.BEmu.CorrelationID;
import java.util.List;
import java.util.ArrayList;
import java.util.Map;
import java.util.HashMap;
import java.util.regex.Pattern;
import java.util.regex.Matcher;

public class EventReference extends Event
{
    public EventReference(RequestReference request) throws Exception
    {
        super._request = request;
        super._messages = this.generateMessages();
    }
    
    private List<Message> generateMessages() throws Exception
    {
        List<Message> result = new ArrayList<Message>();
        RequestReference rreq = (RequestReference)super._request;

        Map<String, Map<String, Object>> securities = new HashMap<String, Map<String, Object>>();
        for(int i = 0; i < rreq.securities().size(); i++)
        {
        	String security = rreq.securities().get(i);
        	if(!securities.containsKey(security))
        	{
        		String strPattern = "[A-Z]{1,4}\\s+\\d{6}[CP]\\d{8} EQUITY";
        		Pattern p = Pattern.compile(strPattern);
        		Matcher m = p.matcher(security.toUpperCase());
        		boolean isOption = m.find();
        		
        		Map<String, Object> fieldData = new HashMap<String, Object>();
        		securities.put(security, fieldData);
        		
        		for(int j = 0; j < rreq.fields().size(); j++)
        		{
        			String field = rreq.fields().get(j);
                    Object value = com.bemu.BEmu.types.RandomDataGenerator.referenceDataFromFieldName(field, security, isOption, rreq);
                    
                    if(value != null)
                    {
                    	fieldData.put(field.toUpperCase(), value);
                    }
        		}
        		
        	}
        }

        CorrelationID corr = super._request.correlationId(); 
        com.bemu.BEmu.ReferenceDataRequest.MessageReference msg = new com.bemu.BEmu.ReferenceDataRequest.MessageReference(corr, securities);
        result.add(msg);
        
        return result;
        
    }
}
